piwik-sdk/src/test/java/org/piwik/sdk/dispatcher/DispatcherTest.java

```diff
@@ -128,6 +133,59 @@ public void testDispatch_gzip() throws Exception {
     }
 
     @Test
+    public void testAfterOptoutSendRemainingDatas() throws Exception {
+        when(tracker.isDryRun()).thenReturn(false);
+        when(tracker.getPiwik()).thenReturn(piwik);
+        when(tracker.getPiwik().isOptOut()).thenReturn(true);
+        dispatcher.setDispatchAfterOptout(true);
+
+        Packet packet = mock(Packet.class);
+
+        URL url = new URL("http://example.com");
+        when(packet.getTargetURL()).thenReturn(url);
+
+        JSONObject jsonObject = new JSONObject();
+        jsonObject.put("test", "test");
+        when(packet.getPostData()).thenReturn(jsonObject);
+
+        HttpURLConnection urlConnection = mock(HttpURLConnection.class);
+        when(packet.openConnection()).thenReturn(urlConnection);
+        OutputStream outputStream = mock(OutputStream.class);
+        when(urlConnection.getOutputStream()).thenReturn(outputStream);
+
+        dispatcher.setDispatchGzipped(false);
+        dispatcher.dispatch(packet);
+        verify(urlConnection, never()).addRequestProperty("Content-Encoding", "gzip");
+
+        dispatcher.setDispatchGzipped(true);
+        dispatcher.dispatch(packet);
+        verify(urlConnection).addRequestProperty("Content-Encoding", "gzip");
+    }
+
+    @Test
+    public void testAfterOptoutLooseRemainingDatas() throws Exception {
+        when(tracker.isDryRun()).thenReturn(false);
+        when(tracker.getPiwik()).thenReturn(piwik);
+        when(tracker.getPiwik().isOptOut()).thenReturn(true);
+        dispatcher.setDispatchAfterOptout(false);
+
+        Packet packet = mock(Packet.class);
```

Comment:
You mock a packet here, but it's just a new mock packet, it won't have any influence on the internal `Packet` objects created by `Dispatcher`.

We need better architecture to test this. I'm working on allowing better testing in #148